src/distributed_join.cuh

```diff
@@ -152,6 +163,219 @@ void all_to_all_comm(cudf::table_view input,
+ * compressed data of a column received from each remote GPUs. This argument does not need to be
+ * preallocated.
+ * @param[out] compressed_recv_offset `compressed_recv_offset[i,j]` represents the start index of
+ * compressed data in `compressed_output[i]` received from rank `j`.
+ * @param[in] communicator An instance of `Communicator` used for communication.
+ * @param[in] include_self If true, this function will send the partition destined to the current
+ * rank.
+ */
+void all_to_all_comm_with_compression(cudf::table_view input,
+                                      vector<cudf::size_type> const &send_offset,
+                                      vector<rmm::device_buffer> &compressed_input,
+                                      vector<rmm::device_buffer> &compressed_output,
+                                      vector<vector<int64_t>> &compressed_recv_offset,
+                                      Communicator *communicator,
+                                      bool include_self = true)
+{
+  cudf::size_type ncols = input.num_columns();
+  compressed_input.resize(ncols);
+  compressed_output.resize(ncols);
+  compressed_recv_offset.resize(ncols);
+
+  int mpi_rank = communicator->mpi_rank;
+  int mpi_size = communicator->mpi_size;
+
+  for (cudf::size_type icol = 0; icol < ncols; icol++) {
+    cudf::column_view column   = input.column(icol);
+    cudf::data_type dtype      = column.type();
+    cudf::size_type dtype_size = cudf::size_of(dtype);
+
+    vector<rmm::device_buffer> compressed_column(mpi_size);
+    vector<size_t> compressed_column_size(mpi_size, 0);
+
+    // compress each partition in the column separately and store the result in *compressed_column*
+    for (int irank = 0; irank < mpi_size; irank++) {
+      if (!include_self && irank == mpi_rank) continue;
+
+      cudf::type_dispatcher(
+        dtype,
+        compression_functor{},
+        static_cast<const void *>(column.head<char>() + send_offset[irank] * dtype_size),
+        send_offset[irank + 1] - send_offset[irank],
+        compressed_column[irank],
+        compressed_column_size[irank]);
+    }
+
+    // calculate and communicate offsets for compressed column
+    vector<int64_t> compressed_send_offset(mpi_size + 1);
+    compressed_send_offset[0] = 0;
+    for (int irank = 0; irank < mpi_size; irank++) {
+      compressed_send_offset[irank + 1] =
+        compressed_send_offset[irank] + compressed_column_size[irank];
+    }
+
+    communicate_sizes(compressed_send_offset, compressed_recv_offset[icol], communicator);
+
+    // merge compressed data of all partitions in *compressed_column* into a single buffer
+    compressed_input[icol] = rmm::device_buffer(compressed_send_offset.back());
+    for (int irank = 0; irank < mpi_size; irank++) {
+      if (!include_self && irank == mpi_rank) continue;
+
+      CUDA_RT_CALL(
+        cudaMemcpy((void *)((char *)compressed_input[icol].data() + compressed_send_offset[irank]),
```

Comment:
Can you remind me why this memory copy is needed? Maybe add some comments.

Reply:
The all-to-all interface works on a single buffer with offsets. Since we don't know the compressed size without actually doing the compression, we cannot pre-allocate this buffer beforehand. So here we compress each partition separately, and once that's done, we can allocate the buffer and copy the compressed data into the buffer.

So far I think have a device-to-device memory copy is a tolerable overhead, compared to the cost of compression. If that's not the case in the future, maybe we can redesign our all-to-all interface.

This part of the code is refactored in a future MR. I will add some more detailed explanation in the current copy to avoid merge conflict.